File: go/785.go
package main

func dfs(cur int, left bool, graph *[][]int, side *[]int) bool {
	if left {
		if (*side)[cur] == 1 {
			return true
		} else if (*side)[cur] == -1 {
			return false
		}
	} else if !left {
		if (*side)[cur] == -1 {
			return true
		} else if (*side)[cur] == 1 {
			return false
		}
	}
	if left {
		(*side)[cur] = 1
	} else {
		(*side)[cur] = -1
	}
	for _, v := range (*graph)[cur] {
		if !dfs(v, !left, graph, side) {
			return false
		}
	}
	return true
}

func isBipartite(graph [][]int) bool {
	side := make([]int, len(graph))
	for i := range graph {
		if side[i] == 0 {
			if !dfs(i, true, &graph, &side) {
				return false
			}
		}
	}
	return true
}

File: python/785.py
from typing import List


class Solution:
    def dfs(self, graph, cur, left, ls, rs):
        if left and cur in rs:
            return False
        if not left and cur in ls:
            return False
        if left and cur in ls:
            return True
        elif not left and cur in rs:
            return True
        if left:
            ls.add(cur)
        else:
            rs.add(cur)
        for i in graph[cur]:
            if not self.dfs(graph, i, not left, ls, rs):
                return False
        return True

    def isBipartite(self, graph: List[List[int]]) -> bool:
        ls, rs = set(), set()
        for i in range(len(graph)):
            if i not in ls and i not in rs:
                if not self.dfs(graph, i, True, ls, rs):
                    return False
        return True
